EIPS/eip-3584.md

```diff
@@ -0,0 +1,131 @@
+* sorted list of 2-tuples of:
+    * transaction index in which the address or any of its storage keys were accessed
+    * sorted list of storage keys which were accessed
+
+```
+access_list                 := [access_list_entry, ...]
+access_list_entry           := [address, storage_keys, accesses_by_txn_index]
+address                     := bytes20
+accesses_by_txn_index       := [txn_index_and_keys, ...]
+txn_index_and_keys          := [txn_index, storage_keys]
+txn_index                   := uint64  # or uint256 or whatever
+storage_keys                := [storage_key, ...]
+storage_key                 := bytes32
+```
+
+Additional sorting rules for the above are that:
+* `access_list` is sorted by the `address`
+* `storage_keys` is sorted
+* `accesses_by_txn_index` is sorted by `txn_index`
+
+Additional validation rules for the above are that:
+* Each unique `address` may only appear at most once in `access_list`
+* Each `storage_key` may only appear at most once in `storage_keys`
+* Each `txn_index` may only appear at most once in `txn_index_and_keys`
+
+All sorting is in increasing order.
+
+#### AccessListRoot
+An `AccessListRoot` is a [urn](https://en.wikipedia.org/wiki/Uniform_Resource_Name) encoding `Hash/Commitment` of the **canonical** `access_list` as well as the construction type (`sha256/kate/merkle/verkle`).
+
+For a very simple start we can do a `SHA256` of the canonical `access_list` and include it in the block header with the following the urn scheme:
+```
+AccessListRoot := "urn:sha256:${ SHA256( access_list ) }"
+```
+
+This AccessListRoot could evolve to `urn:merkle:...` or to `urn:kate:...` or to any other scheme as per requirement.
+
+#### Additional Block Validation
+Full nodes while accepting, executing the new block have an additional validation check that the block’s `AccessListRoot` matches the one calculated from the one generated by executing the block using the construction as defined by the `AccessListRoot` urn.
+
+#### A JSON RPC Endpoint
+Recommended implementation of fetching the block access list of a given block something like 
+```
+eth_BlockAccessList ( BlockNumberOrBlockHashOrAccessListURN, ListOfAddressesOrOffsetOrNull )
+``` 
+which can return paginated `access_list`, paginated at the top address level, filtered by optional addresses filter.
+
+#### Future optional extensions
+We can extend the notion of a block’s `access_list` to:
+```
+access_list := Set[ Address,
+                    List [ AddressWitnesses ],
+                    Set  [ AccessedStorageKey, List [ StorageKeyWitnesses] ],        
+                    Set  [ AccessedInBlockTransactionNumber, List [ AccessedStorageKeys ] ]]
+```
+and then get to define the a *canonical* construction for bulding the fingerprint.
+This will allow an incremental path to partial or full stateless chain protocols, where it would be easy to bundle/request **witnesses** using this `access_list`.
+
```

Comment:
This should be removed or moved to the rationale.  The specification section should limit itself to only the things that must be implemented by someone wanting to adopt this specification.  It may make sense to mention this in the Rationale if certain design choices were made specifically to enable future expansion.  Alternatively, consider writing another EIP that depends on this one which defines next step, or keep such information in another location.

Reply:
moved to rationale.